src/namespace.rs

```diff
@@ -0,0 +1,37 @@
+use std::{borrow::Cow, ops::Deref};
+
+/// The namespace of a storage container. Meant to be constructed from "stringy" types.
+#[derive(Clone, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub struct Ns(Cow<'static, [u8]>);
+
+impl Ns {
+    pub const fn from_static_str(s: &'static str) -> Ns {
+        Ns(Cow::Borrowed(s.as_bytes()))
+    }
+}
+
+impl From<&'static str> for Ns {
+    fn from(s: &'static str) -> Self {
+        Ns(Cow::Borrowed(s.as_bytes()))
+    }
+}
+
+impl From<String> for Ns {
+    fn from(s: String) -> Self {
+        Ns(Cow::Owned(s.into_bytes()))
+    }
+}
+
+impl From<Cow<'static, [u8]>> for Ns {
+    fn from(s: Cow<'static, [u8]>) -> Self {
+        Ns(s)
+    }
+}
+
+impl Deref for Ns {
```

Comment:
I still see calling `Ns` a smart pointer as confusing since managing heap allocations or performing bounds checks is not its purpose, and it doesn't perform any of that directly. That logic could be stretched to call any struct with a `String` field a smart pointer.

I'll just remove `Deref`, but I'll also note my frustration with the way it breaks all callers.

Reply:
According to this https://doc.rust-lang.org/book/ch15-00-smart-pointers.html

If your struct acts as a pointer, and also implements `Deref` and `Drop`, then it's a smart pointer.

Reply:
That would mean the book contradicts itself as it also calls `String` a smart pointer. `String` does not have a custom `Drop` implementation. I guess they're not so strict with this.

Reply:
I guess `String` is somewhat special, with `&str` and `&[u8]` around.

Reply:
String is a smart pointer. It is a 3-component struct that manages a heap allocation.

Reply:
https://doc.rust-lang.org/src/alloc/string.rs.html#365

Maybe it was in older versions of Rust? It seems to delegate all memory management stuff to `Vec`.